Sprite_Sheets, a new system to allow for better species

Kapu1178 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
This is completely stolen from baystation12.

~~Items now have a sprite_sheets var that can hold a dictionary of SPECIES_ID = (dmi path)~~
Bitfield gaming

The long and short of it is this allows for unique sprites to be assigned to items based on the species of the wearer, allowing for breaking the disgusting "human but X" mold of TG species. 

In addition, I will be adding species_restricted, which will allow per-item species restrictions on the wearer.

- [x] Sprite_sheets basic implementation
- [x] ~~In-hand sprite_sheets~~ A slightly better Pixel Offset system is probably better
- [x] Fully debug sprite_sheets
- [x] Species_restricted basic implementation
- [x] Fully debug species_restricted

Update:

Both sprite_sheets and species_restricted have been changed to bitfields instead. 
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Better species good
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
code: Sprite sheets! A new system to allow for better species in the future.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Kapu1178 committed:
Initial Commit

KubeRoot commented on code/game/objects/items.dm:
```diff
@@ -119,7 +119,10 @@ GLOBAL_VAR_INIT(rpg_loot_items, FALSE)
+	///A dictionary of a species and a DMI path to use as an alternative sprite for any given item
+	var/list/sprite_sheets = null
+	///A list of species that the item cannot be worn by
+	var/list/species_restricted = null
```
Is it a good idea for this to be a blacklist? Sounds like this is something where the specific species won't be able to wear *most* items.

KubeRoot replied:
I'm also concerned that this means any item with either species restriction or per-species sprite will have an additional list for each of those *for every such item in existence.* I don't know nearly enough about BYOND to be able to tell if this is going to be an issue, but I feel like it might need to be reworked to use a global dictionary based on typepath.

Kapu1178 replied:
> 
> 
> I'm also concerned that this means any item with either species restriction or per-species sprite will have an additional list for each of those _for every such item in existence._ I don't know nearly enough about BYOND to be able to tell if this is going to be an issue, but I feel like it might need to be reworked to use a global dictionary based on typepath.

Works fine for Baystation.

Kapu1178 replied:
> 
> 
> Is it a good idea for this to be a blacklist? Sounds like this is something where the specific species won't be able to wear _most_ items.

Species_restricted will allow for flexible blacklists on either a parent object or specific clothing items. It will likely primarily be used to blacklist large amount of items by inheritance, due to a lack of sprites.

Kapu1178 replied:
> 
> 
> it working for baystation is the lowest bar in the world. they barely have any memory or performance concerns due to the lack of gameplay.

If theres a better way to do it then do suggest, this is a draft after all.

Kapu1178 replied:
**TO DO BASED ON SUGGESTIONS**

- [x] ~~Look into using an associative string list instead of a dictionary for sprite_sheets~~ I guess bee doesnt have these?
- [x] Make species_restricted into a bitfield instead, based on bodytype

MistakeNot4892 replied:
Original author of species_restricted on Bay chipping in to say that aping this setup is a bad idea, you would be better served by redesigning it from scratch. When I coded it I had no idea how much of a memory concern lists were.

Kapu1178 replied:
> 
> 
> Original author of species_restricted on Bay chipping in to say that aping this setup is a bad idea, you would be better served by redesigning it from scratch. When I coded it I had no idea how much of a memory concern lists were.

Completed. 🙏

PowerfulBacon commented:
God having to sprite something once is bad enough.

triplezeta commented:
i hope you guys will like needing to resprite every. single. clothing item. ever. every single time you want to add a species that doesnt conform to the humanoid body shape


not to mention the sheer asset bloat, and god forbid you want to actually update the clothes sprites in the future

PowerfulBacon commented:
> 
> 
> i hope you guys will like needing to resprite every. single. clothing item. ever. every single time you want to add a species that doesnt conform to the humanoid body shape
> 
> not to mention the sheer asset bloat, and god forbid you want to actually update the clothes sprites in the future

baycoder moment

PowerfulBacon commented:
In all seriousness though, nobody will ever even add a new species to this since they would need to resprite thousands of items.

triplezeta commented:
https://github.com/shiptest-ss13/Shiptest/pull/100 

alternative system, this requires only twelve icons minimum in order to convert clothing to a new body type and i would suggest it over this as it doesnt ask you to resprite every clothes sprite ever

Kapu1178 commented:
> 
> 
> In all seriousness though, nobody will ever even add a new species to this since they would need to resprite thousands of items.

the entire point of species_restricted is so you can ignore items that realistically nobody is going to care about not being able to wear. In addition, theres nothing saying that a new species such as Naevi's Grod would even be able to wear most human clothing in the first place.

Kapu1178 commented:
> 
> 
> [shiptest-ss13/Shiptest#100](https://github.com/shiptest-ss13/Shiptest/pull/100)
> 
> alternative system, this requires only twelve icons minimum in order to convert clothing to a new body type and i would suggest it over this as it doesnt ask you to resprite every clothes sprite ever

This looks promising, I'd like to see how it looks in practice.

Kapu1178 committed:
Basic species_restricted implementation

species_restricted bitfield & bitflags. Handling for it as well. Updated all species datums to contain their ID macro and bodytype bitflag.

Kapu1178 commented:
After some thought, I think a better way to do sprite_sheets would be to use a bitfield aswell, and use a species proc to return DMI data, to greatly reduce memory bloat.

Kapu1178 committed:
Changed sprite sheets to a bitfield. Update Icons now reads the bitfield and calls the specie's get_custom_icons() proc.

Kapu1178 committed:
Got ahead of my self. Fixed errors

ItsTisme commented:
Iam still convinced that this wont work out, its unreasonble to assume "they wont wear it so it wont matter" not only would you have to repreise all job jumpsuits ATLEAST but also any of the donator items, items from the beeshop, items from the clothes vendor. Syndicate agent outfits. VARIUS misc clothing scatterd around maint etc etc etc. I dont think there are many clothing items we "just dont wear"

KubeRoot commented:
> I dont think there are many clothing items we "just dont wear"

The way I see it, this is just a way to be able to progressively add more sprites as necessary, *preventing you* from wearing anything that doesn't have an appropriate sprite yet. If *that* is something you don't like, then, well, I can definitely see why.

The issue for me is that everybody's talking like you **have** to resprite **every single item**, on a system specifically designed so that you **do not** have to do just that.

Kapu1178 commented:
> 
> 
> > I dont think there are many clothing items we "just dont wear"
> 
> The way I see it, this is just a way to be able to progressively add more sprites as necessary, _preventing you_ from wearing anything that doesn't have an appropriate sprite yet. If _that_ is something you don't like, then, well, I can definitely see why.
> 
> The issue for me is that everybody's talking like you **have** to resprite **every single item**, on a system specifically designed so that you **do not** have to do just that.

I have absolutely no fucking idea how people aren't grasping this.

This system is for making new species that may not even require using some slots at all, and is designed so you only would realistically resprite the core items like jumpsuits, helmets, riot armor, etc. 

We do not need to resprite every single fucking gimmick item that is actually equipped once every 100 rounds.

triplezeta commented:
to me its rather iffy gameplaywise to have some things you can and cannot wear based solely on "someone didnt sprite this yet". things like slots just not existing for certain races, sure, but what about things that are similar to a humanoid body type but different enough to warrant new clothes sprites, such as Vox? regardless of gimmick items the number of practical clothing that exists is immense along with the effort required to resprite it all.

Kapu1178 commented:
> 
> 
> to me its rather iffy gameplaywise to have some things you can and cannot wear based solely on "someone didnt sprite this yet". things like slots just not existing for certain races, sure, but what about things that are similar to a humanoid body type but different enough to warrant new clothes sprites, such as Vox? regardless of gimmick items the number of practical clothing that exists is immense along with the effort required to resprite it all.

Im pretty sure the vast majority of players would be completely fine being unable to wear Slim Jeans if it meant we can have an entirely new species that isnt Human But X.

KubeRoot commented:
> to me its rather iffy gameplaywise to have some things you can and cannot wear based solely on "someone didnt sprite this yet". 

Seems to me like you can just explain this away by certain clothing having adjustable straps or buttons that allows you to wear things as other species... and not making species that are literally humanoid but slightly different shape for the sake of fluff. Because at that point you're just making problems for yourself, without this PR you'd just *have to* resprite everything for it whether you like it or not.

Kapu1178 committed:
Debugged all the primary slots

Leaving debug icons and stuff in for now to make my life easier, and so its easier to see whats going on

Kapu1178 committed:
First review ready

Removed debug spritesheets. Added a prototype species-based hand offset system.

Kapu1178 committed:
Updates comments and adds debug species for reference

BrumalLake commented:
a debug species bitflag but no id define for it alright

Kapu1178 commented:
> a debug species bitflag but no id define for it alright

Doesn’t need one. The bitflag is used for sprite sheets, the ID define would be redundant since it isn’t used.

Kapu1178 committed:
Should pass unit tests now

Kapu1178 commented:
For some reason "NONE" isnt allowed for species changesource flags???

Kapu1178 committed:
Fuck you unit test

Sinestia commented on code/__DEFINES/mobs.dm:
```diff
@@ -77,6 +77,32 @@
+#define SPECIES_FELINID			"felinid"
+#define SPECIES_OOZELING		"oozeling"
+#define SPECIES_FLY 			"fly"
+#define SPECIES_SQUID			"squid" //I forgot squids existed lmao
```
sus comment

Kapu1178 replied:
its true!

Naevii commented:
> 
> 
> In all seriousness though, nobody will ever even add a new species to this since they would need to resprite thousands of items.

wrong, we are already doing that!
https://github.com/BeeStation/BeeStation-Hornet/pull/5442

Kapu1178 committed:
Fixes inv slots

Ill write a seethe post about this later. Fucking TG.

francinum commented:
Seems stable after a day's worth of testing.
